### PR TITLE
introduce OperatorContext resource

### PR DESCRIPTION
**Pull request by gernest**
The `OperatorContext`  holds information needed by the operator to talk to gateway admin api

example resource usage
```yaml
apiVersion: tyk.tyk.io/v1alpha1
kind: OperatorContext
metadata:
  name: local-pro-test
spec:
  secretRef:
    namespace: tyk-operator-system
    name: tyk-operator-conf

---
apiVersion: tyk.tyk.io/v1alpha1
kind: ApiDefinition
metadata:
  name: httpbin-with-context
spec:
  contextRef:
    name: local-pro-test
    namespace: default
  name: httpbin-with-context
  use_keyless: true
  protocol: http
  active: true
  proxy:
    target_url: http://httpbin.org
    listen_path: /httpbin-with-context
    strip_listen_path: true
```

In the example above, we are  loading context from secret `tyk-operator-system` . This will load the context from the secret by reading values for these keys in secret

```
TYK_MODE
TYK_URL
TYK_AUTH
TYK_ORG
TYK_TLS_INSECURE_SKIP_VERIFY
```

Or you can directly provide the context values via the env field
```yaml
apiVersion: tyk.tyk.io/v1alpha1
kind: OperatorContext
metadata:
  name: local-pro-test
spec:
  env:
    mode: pro
    auth: TYK_DASH_AUTH_TOKEN
    org: TYK_DASH_ORG_ID
    url: TYK_DASH_URL
    insecureSkipVerify: true
    ingress:
      httpPort: 8000
      httpsPort: 8443
```

### CODE REVIEW

**Comment by asoorm** [comment]
```
apiVersion: tyk.tyk.io/v1alpha1
kind: OperatorContext
metadata:
  name: local-pro-test
spec:
  env:
    mode: pro
    auth: TYK_DASH_AUTH_TOKEN <----------------
    org: TYK_DASH_ORG_ID
    url: TYK_DASH_URL
    insecureSkipVerify: true
    ingress:
      httpPort: 8000
      httpsPort: 8443
```

Is it possible to pull `spec.env.auth: TYK_DASH_AUTH_TOKEN` only from secret?
e.g. using a combination of the two mechamisms, I could override any env values with those found in the secret?

```
apiVersion: tyk.tyk.io/v1alpha1
kind: OperatorContext
metadata:
  name: local-pro-test
spec:
  secretRef:
    namespace: tyk-operator-system
    name: tyk-operator-conf
  env:
    mode: pro
    auth: TYK_DASH_AUTH_TOKEN <----------------
    org: TYK_DASH_ORG_ID
    url: TYK_DASH_URL
    insecureSkipVerify: true
    ingress:
      httpPort: 8000
      httpsPort: 8443
```

File: config/samples/07-operator-context/pro.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+apiVersion: tyk.tyk.io/v1alpha1
+kind: OperatorContext
+metadata:
+  name: local-pro-test
+spec:
+  secretRef:
+    namespace: tyk-operator-system
+    name: tyk-operator-conf
```
Comment (asoorm) [inline config/samples/07-operator-context/pro.yaml:8]: Could provide an example secret resource to illustrate the config options.

Comment (asoorm) [reply config/samples/07-operator-context/pro.yaml:8]: Or in comments:

```
kubectl create secret -n tyk-operator-system generic tyk-operator-conf \
  --from-literal "TYK_AUTH=${TYK_AUTH}" \
  --from-literal "TYK_ORG=${TYK_ORG}" \
  --from-literal "TYK_MODE=${TYK_MODE}" \
  --from-literal "TYK_URL=${TYK_URL}" \
  --from-literal "TYK_TLS_INSECURE_SKIP_VERIFY=true"
```

Something like this

File: config/samples/tyk_v1alpha1_operatorcontext.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+apiVersion: tyk.tyk.io/v1alpha1
+kind: OperatorContext
+metadata:
+  name: operatorcontext-sample
+spec:
+  # Add fields here
+  foo: bar
```
Comment (asoorm) [inline config/samples/tyk_v1alpha1_operatorcontext.yaml:7]: Can we delete these files - and stop the framework scaffolding them?

File: controllers/apidefinition_controller.go [hunk 3]
```diff
@@ -95,6 +95,9 @@ func (r *ApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		if desired.Spec.APIID == "" {
 			desired.Spec.APIID = encodeNS(req.NamespacedName.String())
 		}
+		if desired.Spec.OrgID == "" {
+			desired.Spec.OrgID = r.Env.Org
+		}
```
Comment (asoorm) [inline controllers/apidefinition_controller.go:100]: Sanity check - Does this not mutate the apidefinition spec - causing a reconciliation?

File: controllers/securitypolicy_controller.go [hunk 4]
```diff
@@ -77,6 +77,9 @@ func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 		if policy.Spec.ID == "" {
 			policy.Spec.ID = encodeNS(ns)
 		}
+		if policy.Spec.OrgID == "" {
+			policy.Spec.OrgID = r.Env.Org
+		}
```
Comment (asoorm) [inline controllers/securitypolicy_controller.go:82]: same question as above in apidefinition controller

**Comment by asoorm** [comment]
This is very exciting.

File: controllers/apidefinition_controller.go (hunk 3 above)
Comment (asoorm) [reply controllers/apidefinition_controller.go:100]: e.g. we don't want this to happen again: https://github.com/TykTechnologies/tyk-operator/issues/250

File: config/samples/tyk_v1alpha1_operatorcontext.yaml (hunk 2 above)
Comment (gernest) [reply config/samples/tyk_v1alpha1_operatorcontext.yaml]: done b73d21b

File: config/samples/07-operator-context/pro.yaml (hunk 1 above)
Comment (gernest) [reply config/samples/07-operator-context/pro.yaml:16]: done 6d03096

**Review by asoorm — APPROVED** [review]
Nice

**Pull request merged**